### PR TITLE
Add DATA_SEG and CODE_SEG macro for assigning functions/data to a section

### PR DESCRIPTION
**Pull request by Extravert-ir**
## Purpose
This gives the ability not to only assign functions to `INIT` section, but to `PAGE` and other ones as well

### CODE REVIEW

File: sdk/include/ndk/section_attribs.h [hunk 1]
```diff
@@ -20,19 +20,16 @@ Header Name:
...
-#define INIT_SECTION  __attribute__((section ("INIT")))
-#define INIT_FUNCTION __attribute__((section ("INIT")))
+#define DATA_SEG(segment) __attribute__((section (segment)))
+#define CODE_SEG(segment) __attribute__((section (segment)))
 
 #elif defined(_MSC_VER)
 
 #pragma comment(linker, "/SECTION:INIT,ERW")
-#define INIT_SECTION  __declspec(allocate("INIT"))
-#if (_MSC_VER >= 1800) // Visual Studio 2013 / version 12.0
```
Comment (HBelusca) [inline sdk/include/ndk/section_attribs.h:30]: Great.... really great.... 👿 👿 👿

**Comment by tkreuzer** [comment]
Why does this need to change all instances of INIT_FUNCTION? This seems completely arbitrary and needless.

File: hal/halx86/pcidata.cmake [hunk 2]
```diff
@@ -3,11 +3,11 @@
 #
 add_custom_command(
     OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/pci_classes.c ${CMAKE_CURRENT_BINARY_DIR}/pci_classes.h
-    COMMAND native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_classes.ids ${CMAKE_CURRENT_BINARY_DIR}/pci_classes.c ${CMAKE_CURRENT_BINARY_DIR}/pci_classes.h BINSTR ClassTable INIT_SECTION ${CMAKE_CURRENT_SOURCE_DIR}/include/hal.h
-    DEPENDS native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_classes.ids)
+    COMMAND native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_classes.ids ${CMAKE_CURRENT_BINARY_DIR}/pci_classes.c ${CMAKE_CURRENT_BINARY_DIR}/pci_classes.h BINSTR ClassTable DATA_SEG\(\"INIT\"\) ${CMAKE_CURRENT_SOURCE_DIR}/include/hal.h
```
Comment (katahiromz) [inline hal/halx86/pcidata.cmake:6]: For your information

File: hal/halx86/pcidata.cmake [hunk 3]
```diff
@@ -3,11 +3,11 @@
...
+    DEPENDS native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_classes.ids VERBATIM)
 
 add_custom_command(
     OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/pci_vendors.c ${CMAKE_CURRENT_BINARY_DIR}/pci_vendors.h
-    COMMAND native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_vendors.ids ${CMAKE_CURRENT_BINARY_DIR}/pci_vendors.c ${CMAKE_CURRENT_BINARY_DIR}/pci_vendors.h BINSTR VendorTable INIT_SECTION ${CMAKE_CURRENT_SOURCE_DIR}/include/hal.h
-    DEPENDS native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_vendors.ids)
+    COMMAND native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_vendors.ids ${CMAKE_CURRENT_BINARY_DIR}/pci_vendors.c ${CMAKE_CURRENT_BINARY_DIR}/pci_vendors.h BINSTR VendorTable DATA_SEG\(\"INIT\"\) ${CMAKE_CURRENT_SOURCE_DIR}/include/hal.h
+    DEPENDS native-bin2c ${CMAKE_CURRENT_SOURCE_DIR}/legacy/bus/pci_vendors.ids VERBATIM)
```
Comment (katahiromz) [inline hal/halx86/pcidata.cmake:12]: For your information

**Comment by Extravert-ir** [comment]
After that I'm going to add a bunch of `PAGE` section allocations (and remove most of the `alloc_text` usages). I think it's better to have them done in a uniform way.
The other option would be to introduce `PAGE_FUNCTION`/`PAGE_SECTION` macro

File: sdk/include/ndk/section_attribs.h [hunk 4]
```diff
@@ -20,19 +20,16 @@ Header Name:
 
 #if defined(__GNUC__) || defined(__clang__)
 
-#define INIT_SECTION  __attribute__((section ("INIT")))
-#define INIT_FUNCTION __attribute__((section ("INIT")))
+#define DATA_SEG(segment) __attribute__((section (segment)))
+#define CODE_SEG(segment) __attribute__((section (segment)))
```
Comment (katahiromz) [inline sdk/include/ndk/section_attribs.h:24]: For your information

File: sdk/include/ndk/section_attribs.h [hunk 5]
```diff
@@ -20,19 +20,16 @@ Header Name:
...
-#if (_MSC_VER >= 1800) // Visual Studio 2013 / version 12.0
-#define INIT_FUNCTION __declspec(code_seg("INIT"))
-#else
-#pragma section("INIT", read,execute,discard)
-#define INIT_FUNCTION
-#endif
+#pragma comment(linker, "/SECTION:PAGE,ERW")
+
+#define DATA_SEG(segment) __declspec(allocate(segment))
+#define CODE_SEG(segment) __declspec(code_seg(segment))
```
Comment (katahiromz) [inline sdk/include/ndk/section_attribs.h:32]: For your information

**Review by katahiromz — COMMENTED** [review]
I just marked non-trivial parts.

File: ntoskrnl/include/internal/i386/ke.h [hunk 6]
```diff
@@ -321,7 +321,7 @@ KiSetTebBase(PKPCR Pcr, PVOID TebAddress)
     Ke386SetGdtEntryBase(&Pcr->GDT[KGDT_R3_TEB / sizeof(KGDTENTRY)], TebAddress);
 }
 
-INIT_FUNCTION
+CODE_SEG("INIT")
```
Comment (tkreuzer) [inline ntoskrnl/include/internal/i386/ke.h:324]: I suggest removing them from all prototypes.

**Review by tkreuzer — APPROVED** [review]
Ok, if you think it's better to have this consistent, I'm fine with it.

File: ntoskrnl/io/pnpmgr/plugplay.c [hunk 7]
```diff
@@ -30,7 +30,7 @@ static KEVENT IopPnpNotifyEvent;
 
 /* FUNCTIONS *****************************************************************/
 
-NTSTATUS INIT_FUNCTION
+NTSTATUS CODE_SEG("INIT")
```
Comment (tkreuzer) [inline ntoskrnl/io/pnpmgr/plugplay.c:33]: Should be before the NTSTATUS

**Pull request merged**